sdks/android/Mentat/library/src/androidTest/java/com/mozilla/mentat/FFIIntegrationTest.java

```diff
@@ -1241,5 +1245,77 @@ public void handleRow(TupleResult row) {
         synchronized (expectation2) {
             expectation2.wait(1000);
         }
+        assertTrue(expectation2.isFulfilled);
+    }
+
+    @Test
+    public void testCaching() throws InterruptedException {
+        String query = "[:find ?district :where\n" +
+                "    [?neighborhood :neighborhood/name \"Beacon Hill\"]\n" +
+                "    [?neighborhood :neighborhood/district ?d]\n" +
+                "    [?d :district/name ?district]]";
+
+        Mentat mentat = openAndInitializeCitiesStore();
+
+        final Expectation expectation1 = new Expectation();
+        final QueryTimer uncachedTimer = new QueryTimer();
+        uncachedTimer.start();
+        mentat.query(query).run(new RelResultHandler() {
+            @Override
+            public void handleRows(RelResult rows) {
+                uncachedTimer.end();
+                assertNotNull(rows);
+                expectation1.fulfill();
+            }
+        });
+
+        synchronized (expectation1) {
+            expectation1.wait(1000);
+        }
+        assertTrue(expectation1.isFulfilled);
+
+        mentat.cache(":neighborhood/name", CacheDirection.REVERSE);
+        mentat.cache(":neighborhood/district", CacheDirection.FORWARD);
+
+        final Expectation expectation2 = new Expectation();
+        final QueryTimer cachedTimer = new QueryTimer();
+        cachedTimer.start();
+        mentat.query(query).run(new RelResultHandler() {
+            @Override
+            public void handleRows(RelResult rows) {
+                cachedTimer.end();
+                assertNotNull(rows);
+                expectation2.fulfill();
+            }
+        });
+
+        synchronized (expectation2) {
+            expectation2.wait(1000);
+        }
+        assertTrue(expectation2.isFulfilled);
+
+        long timingDifference = uncachedTimer.duration() - cachedTimer.duration();
+        Log.d("testCaching", "Cached query is "+ timingDifference +" nanoseconds faster than the uncached query");
+
+        assert cachedTimer.duration() < uncachedTimer.duration();
+
+    }
+
+}
+
+class QueryTimer {
```

Comment:
Is this even legal Java?  Two classes at top-level?  Make this an inner class, or (if you think it'll be used) make it an exposed class.  Perhaps you want to make it easier to run `.query()` with a given timer?

Reply:
Two classes are allowed at the top level if only one of them is public. It's just for tests so I didn't think it would matter too much, but I'll make it an inner class to be sure.